v2/tools/generator/internal/codegen/pipeline/export_controller_type_registrations.go

```diff
@@ -71,28 +72,100 @@ func ExportControllerResourceRegistrations(idFactory astmodel.IdentifierFactory,
 }
 
 func handleSecretPropertyChains(
-	chains [][]*astmodel.PropertyDefinition,
+	chains []*propertyChain,
 	idFactory astmodel.IdentifierFactory,
 	def astmodel.TypeDefinition,
 ) ([]*functions.IndexRegistrationFunction, []string) {
 	indexFunctions := make([]*functions.IndexRegistrationFunction, 0, len(chains))
 	secretPropertyKeys := make([]string, 0, len(chains))
 
+	ensureIndexPropertyPathsUnique(chains)
+
 	for _, chain := range chains {
-		secretPropertyKey := makeIndexPropertyKey(chain)
+		secretPropertyKey := chain.indexPropertyKey()
 		indexFunction := functions.NewIndexRegistrationFunction(
-			makeUniqueIndexMethodName(idFactory, def.Name(), chain),
+			chain.indexMethodName(idFactory, def.Name()),
 			def.Name(),
 			secretPropertyKey,
-			chain)
+			chain.properties())
 		indexFunctions = append(indexFunctions, indexFunction)
 		secretPropertyKeys = append(secretPropertyKeys, secretPropertyKey)
 	}
 
 	return indexFunctions, secretPropertyKeys
 }
 
-func catalogSecretPropertyChains(def astmodel.TypeDefinition, definitions astmodel.TypeDefinitionSet) ([][]*astmodel.PropertyDefinition, error) {
+// ensureIndexPropertyPathsUnique looks for conflicting index property paths (which would lead to conflicting index
+// method names) ensures they all produce unique names.
+func ensureIndexPropertyPathsUnique(chains []*propertyChain) {
+	// First mark all the properties at the end of each chain as required
+	for _, chain := range chains {
+		chain.requiredForPropertyPath = true
+	}
+
+	// Look until either we have no collisions, or we can't resolve them
+	for {
+		// Look for collisions
+		collisions := make(map[string][]*propertyChain)
```

Comment:
This isn't actually `collisions`, is it? It contains the names of all chains including changes that didn't collide? Would this be more clear as `proposedNames` or something?

Reply:
Renamed to `chainsByName` as it's acting as an index where we can determine if two (or more) chains are sharing the same name.